utils/src/main/kotlin/com/exactpro/th2/lwdataprovider/MessageSearcher.kt

```diff
@@ -0,0 +1,156 @@
+    private val defaultSessionAlias: String = "",
+) {
+    /**
+     * Searches the first message in [BASE64_FORMAT] format matched by filter. Execute search from [Instant.now] to previous time
+     * @param sessionGroup is parameter for request. Searcher uses [sessionAlias] value when [sessionGroup] is null or blank
+     * @param interval for searching. Searcher stops to send requests when the next `from` time is before than search start time minus [interval]
+     */
+    fun searchLastOrNull(
+        book: String = defaultBook,
+        sessionGroup: String? = defaultSessionGroup,
+        sessionAlias: String = defaultSessionAlias,
+        direction: Direction,
+        interval: Duration,
+        filter: Function<MessageSearchResponse, Boolean>
+    ): MessageSearchResponse? {
+        require(book.isNotBlank()) {
+            "'book' can't be blank"
+        }
+        require(sessionAlias.isNotBlank()) {
+            "'sessionAlias' can't be blank"
+        }
+        require(!interval.isZero) {
+            "'duration' can't be zero"
+        }
+        require(!interval.isNegative) {
+            "'duration' can't be negative"
+        }
+
+        val now = Instant.now()
+        val end = now.minus(interval)
+        var from: Instant
+        var to = now
+
+        val requestBuilder = createRequestBuilder(
+            book = book,
+            sessionGroup = sessionGroup?.ifBlank { sessionAlias } ?: sessionAlias,
+            sessionAlias = sessionAlias,
+            direction = direction
+        )
+        K_LOGGER.info { "Search the last or null, base request: ${requestBuilder.toJson()}, start time: $now, end time: $end" }
+
+        return withCancellation {
+            generateSequence {
+                if (to.isBefore(end)) {
+                    return@generateSequence null
+                }
+
+                from = to
+                to = from.minus(requestInterval)
+
+                service.searchMessageGroups(
+                    requestBuilder.apply {
+                        startTimestamp = from.toTimestamp()
+                        endTimestamp = to.toTimestamp()
+                    }.build().also {
+                        K_LOGGER.debug { "Request from: $from, to: $to" }
+                    }
+                )
+            }.flatMap(Iterator<MessageSearchResponse>::asSequence)
+                .onEach { K_LOGGER.trace { "Message: ${it.toJson()}" } }
+                .firstOrNull(filter::apply)
+        }
+    }
+
+    private fun createRequestBuilder(
+        book: String,
+        sessionGroup: String,
+        sessionAlias: String,
+        direction: Direction,
+    ) = MessageGroupsSearchRequest.newBuilder().apply {
+        bookIdBuilder.name = book
+        addMessageGroup(MessageGroupsSearchRequest.Group.newBuilder().setName(sessionGroup))
+        addStream(
+            MessageStream.newBuilder()
+                .setName(sessionAlias)
+                .setDirection(direction)
+        )
```

Comment:
Do we need to add this if the session alias is empty? And what if we need both directions? Probably, we should add options to request all aliases and both directions (maybe simply don't add the stream filter).

Also, what about the case when we need more than one session alias? The current API does not allow smooth migration to the several session aliases.

Probably, we should change the `searchLastOrNull` to something like this:
```kotlin
    data class MessageStream(val alias: String, val direction: Direction)
    fun searchLastOrNull(
        step: Duration = defaultStep,
        book: String = defaultBook,
        sessionGroup: String = defaultSessionGroup,
        messageStreams: Set<MessageStream> = defaultMessageStreams,
        filter: Function<MessageSearchResponse, Boolean>
    ): MessageSearchResponse?
```